agent/engine/dockerauth/ecr.go

```diff
@@ -17,52 +17,101 @@ import (
 	"encoding/base64"
 	"fmt"
 	"strings"
+	"sync"
+	"time"
 
 	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/async"
 	"github.com/aws/amazon-ecs-agent/agent/ecr"
 	ecrapi "github.com/aws/amazon-ecs-agent/agent/ecr/model/ecr"
+	"github.com/aws/amazon-ecs-agent/agent/utils"
 	"github.com/aws/aws-sdk-go/aws"
 	log "github.com/cihub/seelog"
 	docker "github.com/fsouza/go-dockerclient"
 )
 
+type cacheKey struct {
+	region           string
+	rolearn          string
+	registryID       string
+	endpointOverride string
+}
+
 type ecrAuthProvider struct {
-	authData *api.ECRAuthData
-	client   ecr.ECRClient
+	tokenCache async.Cache
+	factory    ecr.ECRFactory
+	cacheLock  sync.RWMutex
 }
 
-const proxyEndpointScheme = "https://"
+const (
+	roundtripTimeout      = 5 * time.Second
+	tokenCacheSize        = 100
+	MinimumJitterDuration = 30 * time.Minute
+	tokenCacheTTL         = 12 * time.Hour
+	proxyEndpointScheme   = "https://"
+)
+
+func (key *cacheKey) String() string {
+	return fmt.Sprintf("%s-%s-%s-%s", key.rolearn, key.region, key.registryID, key.endpointOverride)
+}
 
 // NewECRAuthProvider returns a DockerAuthProvider that can handle retrieve
 // credentials for pulling from Amazon EC2 Container Registry
-func NewECRAuthProvider(authData *api.ECRAuthData, clientFactory ecr.ECRFactory) DockerAuthProvider {
-	if authData == nil {
-		return &ecrAuthProvider{}
-	}
-	log.Tracef("Getting client in %s with endpoint %s", authData.Region, authData.EndpointOverride)
+func NewECRAuthProvider(ecrFactory ecr.ECRFactory) DockerAuthProvider {
 	return &ecrAuthProvider{
-		authData: authData,
-		client:   clientFactory.GetClient(authData.Region, authData.EndpointOverride),
+		tokenCache: async.NewLRUCache(tokenCacheSize, tokenCacheTTL),
```

Comment:
As per offline discussion, I don't think this gives you the caching behavior that's desired. Can you verify/rework this so that caching behavior for ECR fits the following model?
1. If task has an `executionRole`, all containers in the task share the same credentials auth only once to get creds
2. All tasks that do not have `executionRole` set share the cache for credentials, where the key is based on region + endpoint.